### DISCUSSION TITLE
Can I change points color according to graph values ?

### DISCUSSION THREAD

**Discussion by sylv42240** [patrykandpatrick/vico#482] [Ideas]
Hello, I really like your lib but I was wondering if it possible to change line chart points color according to entries ?
![Capture](https://github.com/patrykandpatrick/vico/assets/50032448/1b1dfc04-8dd4-4b2a-8d36-d76bc8dda97b)
For exemple in this screenhot I'd like to have green points for all the points that are in the green zone, is it possible ?

**Comment by patrickmichalik** [comment]
Hello! This isn’t possible yet, but we’ll be looking into adding such a feature. In some cases, as a workaround, the data can be split into multiple series, one for each color, and multiple `LineSpec`s can be used.

**Reply by SebPawp** [reply]
It's tricky but I was able to do with a custom LineCartesianLayer.LineSpec.point Component and by "reverse calculating" the value based on the point's Y position in the chart bounds.
<img width="305" alt="Screenshot 2024-01-09 at 2 36 08 PM" src="https://github.com/patrykandpatrick/vico/assets/129093932/c8f226cc-e522-43ad-8e70-2106a9f6a34b">

Ideally the Component's `draw` function would take the Entry so we could easily customize the Component based on the value.

In addition, it would also be great if we could have a dynamic label TextComponent to render a `dataLabel` based on the Entry. As part of my workaround, I draw the data label directly inside my custom Point Component.

**Reply by patrickmichalik** [reply]
Thanks for sharing, @SebPawp! The feature in question here is planned, as is improved customization for data labels. `Component` is a low-level API, and it’s used in contexts where there are no chart entries. Thus, `Component#draw` can’t be updated as described. However, the API that we plan on implementing will be specific to line charts and will certainly expose chart entries, enabling you to easily use their _x_ and _y_ values for customization.

**Reply by diegoup2** [reply]
Hello @SebPawp, mind sharing your custom `LineCartesianLayer.LineSpec.point` please.

**Comment by shrift** [comment]
Just want to chime in here that I'm also looking for this feature.
Has this already been implemented in alpha versions that I could start looking at?

**Reply by patrickmichalik** [reply]
Hello! This functionality hasn’t been added yet, but it’s planned for Vico 2, which is currently in alpha. [Vico 2.0.0 Alpha 13](https://github.com/patrykandpatrick/vico/releases/tag/v2.0.0-alpha.13) adds a similar feature for column charts, so we have the general approach figured out.

**Comment by shrift** [comment]
In case it's useful to anyone else, I also hacked around this limitation of points by abusing persistent markers:
```
            val persistentMarkers = remember(entries[0], selectedMeasurement) {
                entries[0].associate { entry ->
                    if (entry.measurement == selectedMeasurement) {
                        when {
                            entry.y >= 100f -> {
                                entry.x to markerSelectedRed
                            }
                            entry.y <= 80f -> {
                                entry.x to markerSelectedWarn
                            }
                            else -> {
                                entry.x to markerSelectedNormal
                            }
                        }
                    } else if (entry.y >= 100f) {
                        entry.x to markerRed
                    } else if (entry.y <= 80f) {
                        entry.x to markerWarn
                    } else {
                        entry.x to markerNormal
                    }
                }
            }
            val markerChangeListener = remember {
                object : MarkerVisibilityChangeListener {

                    override fun onMarkerShown(marker: Marker, markerEntryModels: List<Marker.EntryModel>) {
                        val entry = markerEntryModels.first().entry
                        val measurement = (entry as GGChartEntry).measurement
                        onMeasurementSelected.invoke(
                            measurementType, measurement.time.toInstant().toEpochMilli(), measurement.value
                        )
                    }

                    override fun onMarkerHidden(marker: Marker) {
                        onMeasurementSelected.invoke(measurementType, null, null)
                    }

                    override fun onMarkerMoved(marker: Marker, markerEntryModels: List<Marker.EntryModel>) {
                        val entry = markerEntryModels.first().entry
                        val measurement = (entry as GGChartEntry).measurement
                        onMeasurementSelected.invoke(
                            measurementType, measurement.time.toInstant().toEpochMilli(), measurement.value
                        )
                    }
                }
            }
```
<img width="510" alt="image" src="https://github.com/patrykandpatrick/vico/assets/260590/8bb493d1-d6b4-4276-9043-e902128dcdfe">

This was on 1.14.0. (Sorry for my messy test data making graph look horrid!)

**Reply by patrickmichalik** [reply]
Thanks for sharing!

**Reply by diegoup2** [reply]
For future reference, this only works for 1 line series only, not multiple. Is there a way to assign custom markers for different line series @patrickmichalik ?

**Reply by Gowsky** [reply]
@diegoup2 a `CartesianMarker` highlights everything at a given _x_ value. But it can draw different indicators for different series. However this workaround is no longer needed. [See here](https://github.com/patrykandpatrick/vico/discussions/482#discussioncomment-10257123).

**Comment by patrickmichalik** [comment]
`LineCartesianLayer.PointProvider`, introduced in [Vico 2.0.0 Alpha 23](https://github.com/patrykandpatrick/vico/releases/tag/v2.0.0-alpha.23), makes this possible. Cheers!

**Comment by Gowsky** [comment]
Hello. [Vico 2.0.0](https://github.com/patrykandpatrick/vico/releases/tag/v2.0.0) is now available.

**Comment by Selva-GG** [comment]
how to keep a condition for point connector

**Reply by Gowsky** [reply]
Could you clarify what problem you’re looking to solve?

**Reply by Selva-GG** [reply]
I'm using a scrollable line graph in Vico with timestamps on the X-axis. I want to conditionally break the line connection between points if the time gap between them is more than 2 days.

I tried implementing a custom PointConnector, but it didn’t work as expected

How can I properly implement gap detection between points so that Vico skips drawing a line

**Reply by Gowsky** [reply]
You can achieve this by adding multiple series split by the gaps within `lineSeries` lambda when running a transaction. For example:
```kt
modelProducer.runTransaction {
  lineSeries {
    series(x = listOf(0, 1, 2), y = listOf(3, 2, 2))
    series(x = listOf(3, 4), y = listOf(2, 1))
  }
}
```
This produces the following line chart.
<img src="https://github.com/user-attachments/assets/16e49f08-83e7-4e71-8695-994cb54e4392" width=320/>

**Reply by Selva-GG** [reply]
I actually figured it out earlier, but I was thinking — if the PointConnector supported a lambda condition to decide whether to connect two points, it would make things much easier and cleaner. For large datasets, manually splitting the data can introduce UI delays, so a built-in conditional check would be a nice optimization

**Reply by Selva-GG** [reply]
Anyway, I’ll continue with this approach for now. Thanks for the reply — looking forward to any updates regarding this!

**Reply by Gowsky** [reply]
> For large datasets, manually splitting the data can introduce UI delays

This could be an issue if the transformation were run on the main thread, but it should happen elsewhere. Once off the main thread, it shouldn’t cause any noticeable initialization delays — its complexity is low.
